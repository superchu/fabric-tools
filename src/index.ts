/*
 * Fabric Tools
 */

export interface Doc {
  version: number;
  type: string;
  content: Node[];
}

export interface Node {
  type: string;
  content?: Node[];
}

const filterContent = (content: Node[], ...types: string[]) => {
  return content.reduce((acc, node, index) => {
    if (types.indexOf(node.type) !== -1) {
      if (node.content) {
        acc.push({
          ...node,
          content: filterContent(node.content, ...types)
        });
      } else {
        acc.push(node);
      }
    } else if (node.content) {
      filterContent(node.content, ...types)
        .forEach(child => acc.push(child));
    }

    return acc;
  }, [] as Node[]);
};

export const filterContentByType = (doc: Doc, ...types: string[]) => {
  const { content } = doc;

  if (!content) {
    return [];
  }

  return filterContent(content, ...types);

};
