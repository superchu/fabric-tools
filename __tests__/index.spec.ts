import { filterContentByType } from '../src';
import {
  doc,
  docNested
} from './test-data';

describe('filterContentByType', () => {

  it('should return content filtered by type', () => {
    expect(
      filterContentByType(doc, 'paragraph', 'text')
    ).toMatchSnapshot();
  });

  it('should return empty array if nothing matches', () => {
    expect(
      filterContentByType(doc, 'panel')
    ).toMatchSnapshot();
  });

  it('should be able to extract text from nested document', () => {
    expect(
      filterContentByType(docNested, 'text')
    ).toMatchSnapshot();
  });

  it('should extract valid content only', () => {
    expect(
      filterContentByType(docNested, 'paragraph', 'blockquote', 'text')
    ).toMatchSnapshot();
  });

});
