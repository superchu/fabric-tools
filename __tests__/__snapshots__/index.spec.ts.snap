// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filterContentByType should be able to extract text from nested document 1`] = `
Array [
  Object {
    "text": "One",
    "type": "text",
  },
  Object {
    "text": "Two",
    "type": "text",
  },
  Object {
    "text": "Three",
    "type": "text",
  },
  Object {
    "text": "Four",
    "type": "text",
  },
  Object {
    "text": "Five",
    "type": "text",
  },
]
`;

exports[`filterContentByType should extract valid content only 1`] = `
Array [
  Object {
    "content": Array [
      Object {
        "text": "One",
        "type": "text",
      },
    ],
    "type": "paragraph",
  },
  Object {
    "content": Array [
      Object {
        "text": "Two",
        "type": "text",
      },
    ],
    "type": "paragraph",
  },
  Object {
    "content": Array [
      Object {
        "content": Array [
          Object {
            "text": "Three",
            "type": "text",
          },
        ],
        "type": "paragraph",
      },
      Object {
        "content": Array [
          Object {
            "content": Array [
              Object {
                "text": "Four",
                "type": "text",
              },
            ],
            "type": "paragraph",
          },
        ],
        "type": "blockquote",
      },
      Object {
        "content": Array [
          Object {
            "text": "Five",
            "type": "text",
          },
        ],
        "type": "paragraph",
      },
    ],
    "type": "blockquote",
  },
]
`;

exports[`filterContentByType should return content filtered by type 1`] = `
Array [
  Object {
    "content": Array [
      Object {
        "text": "Hello",
        "type": "text",
      },
    ],
    "type": "paragraph",
  },
  Object {
    "content": Array [
      Object {
        "text": "World",
        "type": "text",
      },
    ],
    "type": "paragraph",
  },
]
`;

exports[`filterContentByType should return empty array if nothing matches 1`] = `Array []`;
